src/Spawner.cpp

```diff
@@ -178,55 +179,31 @@ Spawner::ProcessHandle::ProcessHandle(Deployment *deploment, bool redirectOutput
 
 bool Spawner::ProcessHandle::alive() const
 {
-    //if it was already determined before that the process is already dead,
-    //we can stop here. Otherwise waitpid would fail!
-    if(!isRunning){
-        return isRunning;
-    }
-
     int status = 0;
     pid_t ret = waitpid(pid, &status, WNOHANG);
     
-    if(ret < 0 )
-    {
-        throw std::runtime_error(std::string("WaitPid failed ") + strerror(errno));
-    }
-    
-    if(!status)
-    {
-        return isRunning;
-    }
-    
-    if(WIFEXITED(status))
+    if(ret == 0)
+        return true;
+    else if(ret == pid)
+        return false;
+    else if(ret == -1 )
+        throw std::runtime_error(std::string("waitpid failed: ") + strerror(errno));
+    else
+        throw std::runtime_error("waitpid returned undocumented value");
+}
```

Comment:
I think this function has a bug: the name of the function suggests that it's okay to call it on a non-running process with the expected return value false. This behavior was ensured with the previously present `isRunning` flag.

Whats the behavior now, if we call a process, kill it and then call `alive` after the collection  of the termination signal again. I think the function would not show the expected behavior.

See also comment regarding `killAll` and  `handles`.

Reply:
This should not be an issue. When the process just finished, waitpid will return the pid of the exited process, hence alive will just return false. When waitpid returns -1, it means that it could not check the status of the process.